File: src/Transpiler.Backends/Runtime/values.py
# Typed value operations. Collections and comparer classes remain translated managed IL.
import struct as _struct

class ValueRuntime(HostedRuntime):
    def nullable(self, type_name):
        t = self.meta['types'].get(type_name)
        return t if t and t['definition'] == 'System.Nullable`1' else None

    def nullable_parts(self, value, type_name):
        t = self.nullable(type_name)
        if not t: return None
        has = next(k for k in t['fields'] if self.meta['fields'][k]['owner'] == type_name and k.endswith('::hasValue'))
        data = next(k for k in t['fields'] if self.meta['fields'][k]['owner'] == type_name and k.endswith('::value'))
        return (t['arguments'][0], value.fields[has] != 0, value.fields[data], has, data)

    def box(self, value, type_name):
        parts = self.nullable_parts(value, type_name)
        if parts: return super().box(parts[2], parts[0]) if parts[1] else None
        return super().box(value, type_name)

    def unbox(self, value, type_name, address=False):
        if self.nullable(type_name) and not address:
            result = self.default(type_name)
            parts = self.nullable_parts(result, type_name)
            if value is not None:
                result.fields[parts[3]] = 1
                result.fields[parts[4]] = super().unbox(value, parts[0], False)
            return result
        return super().unbox(value, type_name, address)

    def is_type(self, value, target):
        t = self.nullable(target)
        return super().is_type(value, t['arguments'][0] if t else target)

    def actual_type(self, value, fallback='System.Object'):
        value = self.dereference(value)
        if isinstance(value, CliString): return 'System.String'
        if isinstance(value, CliArray): return value.element + '[]'
        return getattr(value, 'type', fallback)

    def object_slot(self, value, name):
        return self.meta['types'].get(self.actual_type(value), {}).get('objectSlots', {}).get(name)

    def invoke_slot(self, method_id, receiver, rest):
        method = self.meta['methods'][method_id]
        target = self.dereference(receiver)
        if self.meta['types'].get(method['type'], {}).get('valueType'):
            if isinstance(target, CliBox): receiver = self.unbox(target, target.type, True)
            elif not isinstance(receiver, CliRef): receiver = self.cell_ref([target], 0, method['type'])
        else: receiver = target
        return super().call(method_id, [receiver] + rest, False)

    def call(self, method_id, args, virtual=False, constrained=None):
        method = self.meta['methods'][method_id]
        name = {'object.string':'ToString', 'object.hash':'GetHashCode', 'object.equals':'Equals'}.get(method['intrinsic'])
        if virtual and name:
            self.nonnull(self.dereference(args[0]))
            slot = self.object_slot(args[0], name)
            if slot: return self.invoke_slot(slot, args[0], args[1:])
            if name == 'Equals':
                receiver = self.dereference(args[0])
                return self.object_equal(self.box(receiver, receiver.type) if isinstance(receiver, CliValue) else receiver, args[1], False)
        return super().call(method_id, args, virtual, constrained)

    def format(self, value, type_name='System.Object'):
        slot = self.object_slot(value, 'ToString') if value is not None else None
        if slot:
            result = self.invoke_slot(slot, value, [])
            return '' if result is None else self.nonnull(result).text
        return super().format(value, type_name)

    def string_hash(self, text):
        hash_code = 2166136261
        units = text.encode('utf-16-le', 'surrogatepass')
        for i in range(0, len(units), 2):
            hash_code = ((hash_code ^ int.from_bytes(units[i:i+2], 'little')) * 16777619) & 0xffffffff
        return self.bits(hash_code, 32)

    @staticmethod
    def primitive_equal(a, b):
        return a == b or isinstance(a, float) and isinstance(b, float) and math.isnan(a) and math.isnan(b)

    def object_equal(self, a, b, dispatch=True):
        a, b = self.dereference(a), self.dereference(b)
        if a is b: return 1
        if a is None or b is None: return 0
        if dispatch:
            slot = self.object_slot(a, 'Equals')
            if slot: return self.invoke_slot(slot, a, [b])
        if isinstance(a, CliString): return int(isinstance(b, CliString) and a.text == b.text)
        if isinstance(a, CliDelegate): return int(isinstance(b, CliDelegate) and self.delegate_equal(a, b))
        if isinstance(a, CliBox):
            if not isinstance(b, CliBox) or a.type != b.type: return 0
            return self.fields_equal(a.value, b.value, a.type)
        return 0

    def fields_equal(self, a, b, type_name):
        if not isinstance(a, CliValue): return int(self.primitive_equal(a, b))
        if not isinstance(b, CliValue) or a.type != b.type: return 0
        for key in self.meta['types'][type_name]['fields']:
            field = self.meta['fields'][key]
            if field['static']: continue
            if not self.object_equal(self.box(a.fields[key], field['type']), self.box(b.fields[key], field['type'])): return 0
        return 1

    def contract_slot(self, type_name, definition, name):
        if type_name not in self.meta['types']: return None
        contract = definition if definition == 'System.IComparable' else definition + '<' + type_name + '>'
        if not self.assignable(type_name, contract): return None
        for method_id, method in self.meta['methods'].items():
            if method['type'] == contract and method['name'] == name: return method_id
        raise RuntimeError('Semantic interface was not rooted: ' + contract)

    def semantic_equal(self, a, b, type_name):
        pa, pb = self.nullable_parts(a, type_name), self.nullable_parts(b, type_name)
        if pa: return int(pa[1] == pb[1] and (not pa[1] or self.semantic_equal(pa[2], pb[2], pa[0])))
        if a is None or b is None: return int(a is b)
        method = self.contract_slot(type_name, 'System.IEquatable`1', 'Equals')
        if method:
            receiver = self.cell_ref([self.copy_value(a)], 0, type_name) if self.meta['types'].get(type_name, {}).get('valueType') else a
            return self.call(method, [receiver, b], True, type_name)
        return self.object_equal(self.box(a, type_name), self.box(b, type_name))

    def semantic_hash(self, value, type_name, dispatch=True):
        value = self.dereference(value)
        if value is None: return 0
        if dispatch:
            slot = self.object_slot(value, 'GetHashCode')
            if slot: return self.invoke_slot(slot, value, [])
        if isinstance(value, CliBox): return self.semantic_hash(value.value, value.type, False)
        if isinstance(value, CliString): return self.string_hash(value.text)
        nullable = self.nullable_parts(value, type_name)
        if nullable: return self.semantic_hash(nullable[2], nullable[0]) if nullable[1] else 0
        if isinstance(value, float):
            if value == 0 or math.isnan(value): return 2146435072 if math.isnan(value) else 0
            bits = int.from_bytes(_struct.pack('<d', value), 'little')
            return self.bits(bits ^ (bits >> 32), 32)
        if isinstance(value, int):
            if type_name in ('System.Int64', 'System.UInt64'): return self.bits(value ^ (value >> 32), 32)
            if type_name == 'System.Char': return self.bits(value | value << 16, 32)
            return self.bits(value, 32)
        if isinstance(value, CliValue):
            hash_code = 0
            for key in self.meta['types'][value.type]['fields']:
                field = self.meta['fields'][key]
                if not field['static']: hash_code = self.bits(hash_code * 31 + self.semantic_hash(value.fields[key], field['type']), 32)
            return hash_code
        if isinstance(value, CliDelegate):
            hash_code = self.string_hash(value.type)
            for target, method in value.invocations:
                hash_code = self.bits(hash_code * 31 + self.semantic_hash(target, 'System.Object') + self.string_hash(method), 32)
            return hash_code
        return super().external({'intrinsic':'object.identity-hash'}, [value])

    @staticmethod
    def ordinal_compare(a, b):
        if a is b: return 0
        if a is None: return -1
        if b is None: return 1
        x, y = a.text.encode('utf-16-le', 'surrogatepass'), b.text.encode('utf-16-le', 'surrogatepass')
        for i in range(0, min(len(x), len(y)), 2):
            difference = int.from_bytes(x[i:i+2], 'little') - int.from_bytes(y[i:i+2], 'little')
            if difference: return difference
        return (len(x) - len(y)) // 2

    def semantic_compare(self, a, b, type_name):
        na, nb = self.nullable_parts(a, type_name), self.nullable_parts(b, type_name)
        if na: return (self.semantic_compare(na[2], nb[2], na[0]) if nb[1] else 1) if na[1] else (-1 if nb[1] else 0)
        if a is None or b is None: return 0 if a is b else -1 if a is None else 1
        if isinstance(a, (int, float)):
            if isinstance(a, float) and math.isnan(a): return 0 if isinstance(b, float) and math.isnan(b) else -1
            if isinstance(b, float) and math.isnan(b): return 1
            return 0 if a == b else -1 if a < b else 1
        if isinstance(a, CliString):
            raise RuntimeError('TR2300: Culture-dependent string ordering is unavailable; use StringComparer.Ordinal or an explicit comparer.')
        if isinstance(a, CliBox) and isinstance(a.value, (int, float)):
            if not isinstance(b, CliBox) or a.type != b.type: self.fail('System.ArgumentException', 'Objects must have compatible comparison types.')
            return self.semantic_compare(a.value, b.value, a.type)
        method = self.contract_slot(type_name, 'System.IComparable`1', 'CompareTo')
        if method:
            receiver = self.cell_ref([self.copy_value(a)], 0, type_name) if self.meta['types'].get(type_name, {}).get('valueType') else a
            return self.call(method, [receiver, b], True, type_name)
        method = self.contract_slot(self.actual_type(a, type_name), 'System.IComparable', 'CompareTo')
        if method: return self.call(method, [self.box(a, type_name), self.box(b, type_name)], True)
        self.fail('System.ArgumentException', 'At least one object must implement IComparable.')

    def external(self, method, args):
        op = method['intrinsic']
        if op == 'value.equal': return self.semantic_equal(args[0], args[1], method['arguments'][0])
        if op == 'value.hash': return self.semantic_hash(args[0], method['arguments'][0])
        if op == 'value.compare': return self.semantic_compare(args[0], args[1], method['arguments'][0])
        if op == 'value.compare-instance': return self.semantic_compare(self.dereference(args[0]), args[1], method['type'])
        if op == 'string.compare-ordinal': return self.ordinal_compare(args[0], args[1])
        if op == 'object.equals-static': return self.object_equal(args[0], args[1])
        if op == 'object.equals':
            a, b = self.nonnull(self.dereference(args[0])), args[1]
            if method['type'] == 'System.Object': return int(a is b)
            if method['type'] in ('System.ValueType', 'System.Enum'): return self.object_equal(a, b, False)
            if method['type'] == 'System.String': return self.object_equal(a, b, False)
            value = a.value if isinstance(a, CliBox) else a
            if method['params'][0] == 'System.Object': return int(isinstance(b, CliBox) and b.type == method['type'] and self.primitive_equal(value, b.value))
            return self.semantic_equal(value, b, method['type'])
        if op == 'object.hash': return self.semantic_hash(self.nonnull(args[0]), method['type'], False)
        if op == 'object.string':
            value = self.nonnull(self.dereference(args[0]))
            return self.string(super().format(value, value.type if isinstance(value, CliBox) else method['type']))
        return super().external(method, args)

File: src/Transpiler.Backends/Runtime/values.mjs
// Typed value operations and external Object slots. Collection algorithms remain translated managed IL.
class ValueRuntime extends HostedRuntime {
    nullable(type) { const t = this.meta.types[type]; return t?.definition === 'System.Nullable`1' ? t : null; }
    nullable_parts(value, type) {
        const t = this.nullable(type);
        if (!t) return null;
        const has = t.fields.find(k => this.meta.fields[k].owner === type && k.endsWith('::hasValue'));
        const data = t.fields.find(k => this.meta.fields[k].owner === type && k.endsWith('::value'));
        return [t.arguments[0], value.fields[has] !== 0, value.fields[data], has, data];
    }
    box(value, type) {
        const parts = this.nullable_parts(value, type);
        return parts ? parts[1] ? super.box(parts[2], parts[0]) : null : super.box(value, type);
    }
    unbox(value, type, address = false) {
        if (this.nullable(type) && !address) {
            const result = this.default(type), parts = this.nullable_parts(result, type);
            if (value !== null) {
                result.fields[parts[3]] = 1;
                result.fields[parts[4]] = super.unbox(value, parts[0], false);
            }
            return result;
        }
        return super.unbox(value, type, address);
    }
    is_type(value, target) {
        const t = this.nullable(target);
        return t ? super.is_type(value, t.arguments[0]) : super.is_type(value, target);
    }
    actual_type(value, fallback = 'System.Object') {
        value = this.dereference(value);
        if (value instanceof CliString) return 'System.String';
        if (value instanceof CliArray) return value.element + '[]';
        return value?.type ?? fallback;
    }
    object_slot(value, name) { return this.meta.types[this.actual_type(value)]?.objectSlots?.[name] ?? null; }
    invoke_slot(id, receiver, rest) {
        const m = this.meta.methods[id], target = this.dereference(receiver);
        if (this.meta.types[m.type]?.valueType) {
            if (target instanceof CliBox) receiver = this.unbox(target, target.type, true);
            else if (!(receiver instanceof CliRef)) receiver = this.cell_ref([target], 0, m.type);
        } else receiver = target;
        return super.call(id, [receiver, ...rest], false);
    }
    call(id, args, virtual = false, constrained = null) {
        const m = this.meta.methods[id];
        const name = {'object.string': 'ToString', 'object.hash': 'GetHashCode', 'object.equals': 'Equals'}[m.intrinsic];
        if (virtual && name) {
            this.nonnull(this.dereference(args[0]));
            const slot = this.object_slot(args[0], name);
            if (slot) return this.invoke_slot(slot, args[0], args.slice(1));
            if (name === 'Equals') {
                const receiver = this.dereference(args[0]);
                return this.object_equal(receiver instanceof CliValue ? this.box(receiver, receiver.type) : receiver, args[1], false);
            }
        }
        return super.call(id, args, virtual, constrained);
    }
    format(value, type = 'System.Object') {
        const slot = value == null ? null : this.object_slot(value, 'ToString');
        if (slot) {
            const result = this.invoke_slot(slot, value, []);
            return result === null ? '' : this.nonnull(result).text;
        }
        return super.format(value, type);
    }
    // String hashes are runtime-local by contract. Hash UTF-16 code units, not Unicode scalar values.
    string_hash(text) {
        let hash = 2166136261;
        for (let i = 0; i < text.length; i++) hash = Math.imul(hash ^ text.charCodeAt(i), 16777619);
        return hash | 0;
    }
    primitive_equal(a, b) { return a === b || typeof a === 'number' && Number.isNaN(a) && Number.isNaN(b); }
    object_equal(a, b, dispatch = true) {
        a = this.dereference(a); b = this.dereference(b);
        if (a === b) return 1;
        if (a === null || b === null) return 0;
        if (dispatch) { const slot = this.object_slot(a, 'Equals'); if (slot) return this.invoke_slot(slot, a, [b]); }
        if (a instanceof CliString) return Number(b instanceof CliString && a.text === b.text);
        if (a instanceof CliDelegate) return Number(b instanceof CliDelegate && this.delegate_equal(a, b));
        if (a instanceof CliBox) {
            if (!(b instanceof CliBox) || a.type !== b.type) return 0;
            return this.fields_equal(a.value, b.value, a.type);
        }
        return 0;
    }
    fields_equal(a, b, type) {
        if (!(a instanceof CliValue)) return Number(this.primitive_equal(a, b));
        if (!(b instanceof CliValue) || a.type !== b.type) return 0;
        for (const key of this.meta.types[type].fields) {
            const field = this.meta.fields[key]; if (field.static) continue;
            const x = this.box(a.fields[key], field.type), y = this.box(b.fields[key], field.type);
            if (!this.object_equal(x, y)) return 0;
        }
        return 1;
    }
    contract_slot(type, definition, name) {
        const descriptor = this.meta.types[type];
        if (!descriptor) return null;
        const contract = definition === 'System.IComparable' ? definition : definition + '<' + type + '>';
        if (!this.assignable(type, contract)) return null;
        for (const [id, m] of Object.entries(this.meta.methods))
            if (m.type === contract && m.name === name) return id;
        throw new Error('Semantic interface was not rooted: ' + contract);
    }
    semantic_equal(a, b, type) {
        const partsA = this.nullable_parts(a, type), partsB = this.nullable_parts(b, type);
        if (partsA) return Number(partsA[1] === partsB[1] && (!partsA[1] || this.semantic_equal(partsA[2], partsB[2], partsA[0])));
        if (a === null || b === null) return Number(a === b);
        const id = this.contract_slot(type, 'System.IEquatable`1', 'Equals');
        if (id) {
            // Dispatch uses the declared T's contract, not an interface found only on a runtime subtype.
            let receiver = a;
            if (this.meta.types[type]?.valueType) receiver = this.cell_ref([this.copy_value(a)], 0, type);
            return this.call(id, [receiver, b], true, type);
        }
        return this.object_equal(this.box(a, type), this.box(b, type));
    }
    semantic_hash(value, type, dispatch = true) {
        value = this.dereference(value);
        if (value === null) return 0;
        if (dispatch) { const slot = this.object_slot(value, 'GetHashCode'); if (slot) return this.invoke_slot(slot, value, []); }
        if (value instanceof CliBox) return this.semantic_hash(value.value, value.type, false);
        if (value instanceof CliString) return this.string_hash(value.text);
        const nullable = this.nullable_parts(value, type);
        if (nullable) return nullable[1] ? this.semantic_hash(nullable[2], nullable[0]) : 0;
        if (typeof value === 'bigint') return Number(BigInt.asIntN(32, value ^ (value >> 32n)));
        if (typeof value === 'number') {
            if (type === 'System.Double' || type === 'System.Single') {
                if (value === 0 || Number.isNaN(value)) return Number.isNaN(value) ? 2146435072 : 0;
                const buffer = new ArrayBuffer(8), view = new DataView(buffer);
                view.setFloat64(0, value, true);
                return view.getInt32(0, true) ^ view.getInt32(4, true);
            }
            if (type === 'System.Char') return (value | (value << 16)) | 0;
            return value | 0;
        }
        if (value instanceof CliValue) {
            let hash = 0;
            for (const key of this.meta.types[value.type].fields) {
                const field = this.meta.fields[key];
                if (!field.static) hash = (Math.imul(hash, 31) + this.semantic_hash(value.fields[key], field.type)) | 0;
            }
            return hash;
        }
        if (value instanceof CliDelegate) {
            let hash = this.string_hash(value.type);
            for (const [target, method] of value.invocations)
                hash = (Math.imul(hash, 31) + this.semantic_hash(target, 'System.Object') + this.string_hash(method)) | 0;
            return hash;
        }
        return super.external({intrinsic:'object.identity-hash'}, [value]);
    }
    ordinal_compare(a, b) {
        if (a === b) return 0; if (a === null) return -1; if (b === null) return 1;
        const length = Math.min(a.text.length, b.text.length);
        for (let i = 0; i < length; i++) { const d = a.text.charCodeAt(i) - b.text.charCodeAt(i); if (d) return d; }
        return a.text.length - b.text.length;
    }
    semantic_compare(a, b, type) {
        const na = this.nullable_parts(a, type), nb = this.nullable_parts(b, type);
        if (na) return na[1] ? nb[1] ? this.semantic_compare(na[2], nb[2], na[0]) : 1 : nb[1] ? -1 : 0;
        if (a === null || b === null) return a === b ? 0 : a === null ? -1 : 1;
        if (typeof a === 'number' || typeof a === 'bigint') {
            if (Number.isNaN(a)) return Number.isNaN(b) ? 0 : -1;
            if (Number.isNaN(b)) return 1;
            return a === b ? 0 : a < b ? -1 : 1;
        }
        let id = this.contract_slot(type, 'System.IComparable`1', 'CompareTo');
        if (id) {
            let receiver = this.meta.types[type]?.valueType ? this.cell_ref([this.copy_value(a)], 0, type) : a;
            return this.call(id, [receiver, b], true, type);
        }
        id = this.contract_slot(type, 'System.IComparable', 'CompareTo');
        if (id) return this.call(id, [this.box(a,type), this.box(b,type)], true);
        this.fail('System.ArgumentException', 'At least one object must implement IComparable.');
    }
    external(m, args) {
        const op = m.intrinsic;
        if (op === 'value.equal') return this.semantic_equal(args[0], args[1], m.arguments[0]);
        if (op === 'value.hash') return this.semantic_hash(args[0], m.arguments[0]);
        if (op === 'value.compare') return this.semantic_compare(args[0], args[1], m.arguments[0]);
        if (op === 'value.compare-instance') return this.semantic_compare(this.dereference(args[0]), args[1], m.type);
        if (op === 'string.compare-ordinal') return this.ordinal_compare(args[0], args[1]);
        if (op === 'object.equals-static') return this.object_equal(args[0], args[1]);
        if (op === 'object.equals') {
            const a = this.nonnull(this.dereference(args[0])), b = args[1];
            if (m.type === 'System.Object') return Number(a === b); // A direct base.Equals call must not redispatch.
            if (m.type === 'System.ValueType' || m.type === 'System.Enum') return this.object_equal(a, b, false);
            if (m.type === 'System.String') return this.object_equal(a, b, false);
            const value = a instanceof CliBox ? a.value : a;
            if (m.params[0] === 'System.Object') return Number(b instanceof CliBox && b.type === m.type && this.primitive_equal(value, b.value));
            return this.semantic_equal(value, b, m.type);
        }
        if (op === 'object.hash') return this.semantic_hash(this.nonnull(args[0]), m.type, false);
        if (op === 'object.string') {
            // Nonvirtual base calls bypass the override bridge, including inside an override itself.
            const value = this.nonnull(this.dereference(args[0]));
            return this.string(super.format(value, value instanceof CliBox ? value.type : m.type));
        }
        return super.external(m, args);
    }
}
